server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1349,10 +1349,30 @@ public Metadata build() {
 
         private SortedMap<String, IndexAbstraction> buildIndicesLookup() {
             SortedMap<String, IndexAbstraction> indicesLookup = new TreeMap<>();
+            Map<String, DataStream> indexToDataStreamLookup = new HashMap<>();
+            DataStreamMetadata dataStreamMetadata = (DataStreamMetadata) this.customs.get(DataStreamMetadata.TYPE);
+            if (dataStreamMetadata != null) {
+                for (DataStream ds : dataStreamMetadata.dataStreams().values()) {
+                    for (Index i : ds.getIndices()) {
+                        indexToDataStreamLookup.put(i.getName(), ds);
+                    }
+                }
+            }
+
             for (ObjectCursor<IndexMetadata> cursor : indices.values()) {
                 IndexMetadata indexMetadata = cursor.value;
-                IndexAbstraction existing =
-                    indicesLookup.put(indexMetadata.getIndex().getName(), new IndexAbstraction.Index(indexMetadata));
+
+                IndexAbstraction.Index index;
+                DataStream parent = indexToDataStreamLookup.get(indexMetadata.getIndex().getName());
+                if (parent != null) {
+                    List<IndexMetadata> backingIndices = parent.getIndices().stream()
+                        .map(imd -> indices.get(imd.getName())).collect(Collectors.toList());
+                    index = new IndexAbstraction.Index(indexMetadata, new IndexAbstraction.DataStream(parent, backingIndices));
```

Comment:
Can we reuse the `IndexAbstraction.DataStream` created here in the logic below here that adds data streams to the indices lookup? I think it would nice that if a data stream is lookup directly or via a backing index that the same instance is returned.

I think we can do this if we add the data streams to the indicesLookup map before we add the indices to the indicesLookup map?

Reply:
@martijnvg, I agree that would be nice. Adding data streams to `indicesLookup` before indices means that name conflicts will be reported as `Index [foo] conflicts with existing data stream [foo]...` rather than `Data stream [foo] conflicts with existing index [foo]...`. If you are ok with that, I'll make the change.

Reply:
I think that is ok. 

Maybe rename the exception to: `index [foo] and data stream [foo] have the same name`
or something like that, so the name conflict is bad regardless which one was added first.

Reply:
Actually, for consistency and simplicity, I should probably add data streams to the duplicate name check for aliases and indices that occurs before the call to `buildIndicesLookup`.

Reply:
> I should probably add data streams to the duplicate name check for aliases and indices that occurs before the call to buildIndicesLookup.

👍

Reply:
I think it is a requirement that we reuse the same instances, since equals is not overridden.